actor/v7action/service_instance_test.go

```diff
@@ -1233,4 +1233,191 @@ var _ = Describe("Service Instance Actions", func() {
+		)
+
+		var (
+			err      error
+			warnings Warnings
+		)
+
+		JustBeforeEach(func() {
+			warnings, err = actor.UpgradeServiceInstance(fakeServiceInstanceName, fakeSpaceGUID)
+		})
+
+		It("makes a request to get the service instance", func() {
+			Expect(fakeCloudControllerClient.GetServiceInstanceByNameAndSpaceCallCount()).To(Equal(1))
+			actualName, actualSpace, actualQuery := fakeCloudControllerClient.GetServiceInstanceByNameAndSpaceArgsForCall(0)
+			Expect(actualName).To(Equal(fakeServiceInstanceName))
+			Expect(actualSpace).To(Equal(fakeSpaceGUID))
+			Expect(actualQuery).To(BeEmpty())
+		})
+
+		When("the service instance does not exist", func() {
+			BeforeEach(func() {
+				fakeCloudControllerClient.GetServiceInstanceByNameAndSpaceReturns(
+					resources.ServiceInstance{},
+					ccv3.IncludedResources{},
+					ccv3.Warnings{"get SI warning"},
+					ccerror.ServiceInstanceNotFoundError{Name: fakeServiceInstanceName},
+				)
+			})
+
+			It("does not try to upgrade", func() {
+				Expect(fakeCloudControllerClient.UpdateServiceInstanceCallCount()).To(BeZero())
+			})
+
+			It("returns the appropriate error", func() {
+				Expect(warnings).To(ConsistOf("get SI warning"))
+				Expect(err).To(MatchError(actionerror.ServiceInstanceNotFoundError{
+					Name: fakeServiceInstanceName,
+				}))
+			})
+		})
+
+		When("there's no upgrade available", func() {
+			BeforeEach(func() {
+				fakeCloudControllerClient.GetServiceInstanceByNameAndSpaceReturns(
+					resources.ServiceInstance{
+						GUID: "some-guid",
+						Name: "some-name",
+						UpgradeAvailable: types.OptionalBoolean{
+							IsSet: true,
+							Value: false,
+						},
+					},
+					ccv3.IncludedResources{},
+					ccv3.Warnings{"get SI warning"},
+					nil,
+				)
+			})
+
+			It("does not try to upgrade", func() {
+				Expect(fakeCloudControllerClient.UpdateServiceInstanceCallCount()).To(BeZero())
+			})
+
+			It("returns the appropriate error", func() {
+				Expect(warnings).To(ConsistOf("get SI warning"))
+				Expect(err).To(MatchError(actionerror.ServiceInstanceUpgradeNotAvailableError{}))
+			})
+		})
+
+		When("there is an upgrade available", func() {
+			const guid = "some-guid"
+			const planGUID = "some-plan-guid"
+
+			BeforeEach(func() {
+				fakeCloudControllerClient.GetServiceInstanceByNameAndSpaceReturns(
+					resources.ServiceInstance{
+						Type: resources.ManagedServiceInstance,
+						GUID: guid,
+						Name: "some-name",
+						UpgradeAvailable: types.OptionalBoolean{
+							IsSet: true,
+							Value: true,
+						},
+						ServicePlanGUID: planGUID,
+					},
+					ccv3.IncludedResources{},
+					ccv3.Warnings{"warning from get"},
+					nil,
+				)
+
+				fakeCloudControllerClient.GetServicePlanByGUIDReturns(
+					resources.ServicePlan{
+						MaintenanceInfoVersion: "9.1.2",
+					},
+					ccv3.Warnings{"warning from plan"},
+					nil,
+				)
+			})
+
+			Context("and it is done synchronously", func() {
```

Comment:
As of implemented today, you're right, it can't happen.

When implementing/testing it, I saw three options:

1. Consider it might actually return synchronously
2. Return an error if no job url is returned
3. Don't error but also don't test the synchronous case

I went with (1). In the very unlikely scenario we change the API implementation, the CLI won't need to change. I can see how from the documentation standpoint this test might confuse people looking to understand how the API works...

Maybe it's worth going with (3)? 🤔

Reply:
Checking some other methods, like the `CreateServiceBroker`, that seems to be the case: we don't error but we also don't test when no job url is returned... I'll just remove the test